### PR TITLE
A tool to analyze FEP calculations added

### PR DESCRIPTION
**Pull request by tsenapathi**

### CODE REVIEW

**Comment by simonbray** [comment]
This looks good @tsenapathi. I just proposed some small changes. :)

However I had problems with testing the tool. The provided test-data works fine, but when I submit my own data (that I generated with the alchemical_run tool, just a tarfile with 3 xvg files inside) I get an error like this:
```
Plotting the free energy breakdown figure...
Plotting the TI figure...
Traceback (most recent call last):
  File "/home/simon/miniconda3/envs/__alchemical-analysis@1.0.2/bin/alchemical_analysis", line 10, in <module>
    sys.exit(main())
  File "/home/simon/miniconda3/envs/__alchemical-analysis@1.0.2/lib/python2.7/site-packages/alchemical_analysis/alchemical_analysis.py", line 1285, in main
    plotdFvsLambda()
  File "/home/simon/miniconda3/envs/__alchemical-analysis@1.0.2/lib/python2.7/site-packages/alchemical_analysis/alchemical_analysis.py", line 1068, in plotdFvsLambda
    plotTI()
  File "/home/simon/miniconda3/envs/__alchemical-analysis@1.0.2/lib/python2.7/site-packages/alchemical_analysis/alchemical_analysis.py", line 1029, in plotTI
    xt = [i if (i in getInd()) else '' for i in range(K)]
TypeError: argument of type 'NoneType' is not iterable
```
Do you have any ideas?

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml [hunk 1]
```diff
@@ -0,0 +1,95 @@
+<tool id="alchemical_analysis" name="Alchemical Analysis" version="1.0.2">
+    <description>Analysis of alchemical free energy calculations</description> 
+    <requirements>
+         <requirement type="package" version="1.0.2">alchemical-analysis</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+       ln -s '$datain' ./data.tar &&
+       tar xf ./data.tar &&
```
[the remark this answers is not in the document]
Comment (tsenapathi) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: I need a directory of XVG files for all the free energy windows. Is there any better way to upload a directory?

**Comment by tsenapathi** [comment]
How long you have run the simulation? should be at least 100 ps.

**Comment by simonbray** [comment]
Ah, that makes sense, my simulations are only 1 ps. I'll try and generate some longer ones.

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+This tool can run analyze the alchemical free energy simulations.
+
+_____
+
+
+.. class:: infomark
+
+**Input**
+
+       - TAR archive file of the GROMACS XVG files from free energy simulations.
```
Comment (simonbray) [inline tools/free_energy/alchemical_analysis/alchemical_analysis.xml]:
```suggestion
       - TAR archive file of the GROMACS XVG files from free energy simulations. Note that simulations should be at least 100 ps in length.
```

**Comment by tsenapathi** [comment]
@simonbray @bgruening @chrisbarnettster any more comments or suggestions?

**Comment by simonbray** [comment]
Not from me

**Comment by simonbray** [comment]
Actually, wait, a couple more things occurred to me:

- can you add the tool.yml file
- can you make the test data smaller? 36 MB is very big, it should be under 1 MB ideally.

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bgruening) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: About how many files are we talking here? Yes tarballs can be an issue and should be avoided. Galaxy supports uploading hundrets of files and put them into a collection.

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+       tar xf ./data.tar && 
+       mkdir ./RESULTS/ &&
+            #if $lambda.skip_lambda == "no"
+                alchemical_analysis -d data -o ./RESULTS -t '$temp' -s '$time' -c -q xvg -p '$prefix' -g -e -f 10 -w -u '$units' &>> '$output'
+            #else if $lambda.skip_lambda == "yes"
+                 alchemical_analysis -d data -o ./RESULTS -t '$temp' -s '$time' -k '$lambda.skiplambda' -c -q xvg -p '$prefix' -g -e -f 10 -w -u '$units' &>> '$output'
+            #end if
+    ]]></command>
+    <inputs>
+    <param format="tar" name="datain" type="data" label="TI/FEP data output"/>
```
Comment (bgruening) [inline tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: a collection of files?

multiple="True"

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    </param>
+    <conditional name="lambda">
+        <param name="skip_lambda" type="select" label="Skip some lambda states?"  >
+            <option value="no">No</option>
+            <option value="yes">Yes</option>
+        </param>
+            <when value="no">
+            </when>
+            <when value="yes">
+                <param name="skiplambda" type="text" value="1-2-3" label="Indices of lambda states to skip" help="Give lambda indices separated by '-' and they will be removed from the analysis."/>
```
Comment (bgruening) [inline tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: Why is the conditional here needed? If the value is empty and optional="true" then the conditional is not needed, isn't it?

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonbray) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: The number is user-specified but would probably not be as many as hundreds.

Also the order of the files is important. I suppose this is also possible with a collection.

We would also need to update the tool that was just merged (which creates these xvg files) to output a collection.

**Comment by tsenapathi** [comment]
@simonbray smaller data set added 
Can you point me to an example tool.yml file? and why we need to add that?

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml (hunk 3 above)
Comment (tsenapathi) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: @bgruening @simonbray The tools need one directory with all the XVG files and I need to explicitly specify the path to that file and also uploading hundreds of files maybe not convenient, user errors can happen and also histories will look very complex. Generally, we run multiple simulations for one ligand until we get the proper convergence. That's why I thought to use a tar file and keep the inputs simple. If this is not a good idea we can change the input file format.

Comment (bgruening) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: It's really a matter of the amount of data, if its hundreds of files or even thousands, let's go for a tarball at the moment. We could also think about to make it a proper filetype. xvg.tar or something like that. There is no usecase where a user wants to "view"/observe one given file (in such a case)?

In general, I think Galaxy should support the use-case of tar -> collection during upload, but this is another story.

Comment (tsenapathi) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: No no need to view each and every file. Making an xvg.tar file type is a great idea. I will think about that.

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml (hunk 4 above)
Comment (tsenapathi) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: @bgruening I wanted not to show this option by default If wrong values are specified the tool will not run. Should I show this option by default and make it blank?

Comment (bgruening) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: What are correct values and what are wrong, is there a patter that you can check? Galaxy supports "validators", to validate a user-input.

Comment (simonbray) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: @tsenapathi, see in my PR for an example: https://github.com/galaxycomputationalchemistry/galaxy-tools-compchem/pull/44/files#diff-a4db2c2bf23fc0cec5379f8b9d9a92d6R21

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml (hunk 3 above)
Comment (simonbray) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: Personally I think a .xvg.tar filetype is overkill. Packing xvg files into tarballs is not a general usecase IMO.

I also doubt there will be hundreds of xvg files present. When I did TI simulations I had 10 or 20 data points. I guess more than 50 would be quite unusual. So from that point of view a collection should be quite feasible.

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml (hunk 4 above)
Comment (tsenapathi) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: I see. However, if I want to use the argument for skipping lambda (which is -k) I need to specify some parameters, white space is not allowed in my case. That's why I had to use two separate options.

Comment (simonbray) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: You can can change the validator to accept the input you want. See: https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-inputs-param-validator

**Comment by tsenapathi** [comment]
@bgruening @chrisbarnettster @simonbray The condition for skip lambda removed and tried to add a condition based on whether this parameter is given or not. Did not work!! any idea how to do this?

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml [hunk 5]
```diff
@@ -0,0 +1,85 @@
+<tool id="alchemical_analysis" name="Alchemical Analysis" version="1.0.2">
+    <description>Analysis of alchemical free energy calculations</description> 
+    <requirements>
+         <requirement type="package" version="1.0.2">alchemical-analysis</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+       ln -s '$datain' ./data.tar &&
+       tar xf ./data.tar && 
+       mkdir ./RESULTS/ &&
+            #if $skiplambda = "":
```
Comment (simonbray) [inline tools/free_energy/alchemical_analysis/alchemical_analysis.xml]:
```suggestion
            #if $skiplambda == "":
```

Comment (simonbray) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: @tsenapathi, there is also a test file missing

Comment (tsenapathi) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: Fixed it. Thanks @simonbray !!

**Comment by tsenapathi** [comment]
@bgruening @simonbray Good to merge?

**Comment by simonbray** [comment]
There are some things that could still be added:
- a validator to the skiplambda parameter
- we did not reach a decision on this collection/tarball issue
- you need a .shed.yml file, right?

**Comment by tsenapathi** [comment]
@simonbray I have a .shed.yml  see https://github.com/tsenapathi/galaxy-tools-compchem/blob/master/tools/free_energy/.shed.yml.
@bgruening agreed to go with .tar files for now, and in future, I will create a new data type such as xvg.tar.

I will add a validator now

**Comment by simonbray** [comment]
Okay. But please don't make an xvg.tar datatype. A collection is the correct solution longterm.

File: tools/free_energy/alchemical_analysis/alchemical_analysis.xml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    <param name="prefix" type="text" value="md" label="Prefix for datafile sets" help="This should be md if you used the alchemical simulator tool"/>
+    <param name="temp" type="integer" value="300" label="Temperature (K)"/>
+    <param name="time" type="integer" value="0" label="Equilibration time (ps)" help="Discard data prior to this specified time as 'equilibration' data." />
+    <param name="units" type="select" label="Units to report energies" >
+        <option value="kcal">kcal</option>
+        <option value="kJ">kJ</option>
+        <option value="kBT">kT</option>
+    </param>
+    <param name="skiplambda" type="text" label="Indices of lambda states to skip" help="Give a string of lambda indices separated by '-' (e.g. 2-3) and they will be removed from the analysis.  Default: will use all the windows.">
+        <validator type="regex" message="No whitespace allowed">^[ ]*$</validator>
```
Comment (simonbray) [inline tools/free_energy/alchemical_analysis/alchemical_analysis.xml]:
```suggestion
        <validator type="regex" message="Only numeric values and '-' allowed">^[0-9-]*$</validator>
```

Comment (simonbray) [reply tools/free_energy/alchemical_analysis/alchemical_analysis.xml]: try this

**Comment by tsenapathi** [comment]
@bgruening @simonbray can I push the updated gmx_run tool to this PR

**Comment by simonbray** [comment]
@tsenapathi, yes sounds good

**Comment by tsenapathi** [comment]
@simonbray The alchemical analysis test passing on my local installation but keep failing here. Any idea why?

**Comment by simonbray** [comment]
I'll try testing locally

**Comment by simonbray** [comment]
@tsenapathi, my best guess is that my change to the travis config is causing this. Tests are also passing locally for me. I'm reverting it here: https://github.com/galaxycomputationalchemistry/galaxy-tools-compchem/pull/48

**Comment by tsenapathi** [comment]
Cool Thanks!!

**Comment by tsenapathi** [comment]
@simonbray @chrisbarnettster @bgruening any idea why this is happening?

**Comment by tsenapathi** [comment]
@simonbray great! So it needed pymbar.

**Comment by simonbray** [comment]
Nice, it worked! :D 

@tsenapathi, the culprit was the new version of pymbar that was released on conda three days ago. It has some incompatibility with alchemical_analysis. As conda always installs the newest version automatically it's been failing since then.

**Comment by simonbray** [comment]
@bgruening, we can merge afaic. Any other comments on your side?

**Pull request merged**